# See http://lua-users.org/wiki/LuaPowerPatches
# Better signal handling in the interpreter on POSIX systems
# URL: http://lua-users.org/files/wiki_insecure/power_patches/5.1/sig_catch.patch
diff -Nur lua-5.1.5/src/lua.c lua-5.1.5-sig_catch/src/lua.c
--- lua-5.1.5/src/lua.c	2007-12-28 15:32:23.000000000 +0000
+++ lua-5.1.5-sig_catch/src/lua.c	2012-02-18 18:43:08.577043483 +0000
@@ -33,8 +33,6 @@
 
 
 static void laction (int i) {
-  signal(i, SIG_DFL); /* if another SIGINT happens before lstop,
-                              terminate process (default action) */
   lua_sethook(globalL, lstop, LUA_MASKCALL | LUA_MASKRET | LUA_MASKCOUNT, 1);
 }
 
@@ -93,14 +91,24 @@
 }
 
 
+static void sig_catch(int sig, void (*handler)(int))
+{
+  struct sigaction sa;
+  sa.sa_handler = handler;
+  sa.sa_flags = 0;
+  sigemptyset(&sa.sa_mask);
+  sigaction(sig, &sa, 0);         /* XXX ignores errors */
+}
+
+
 static int docall (lua_State *L, int narg, int clear) {
   int status;
   int base = lua_gettop(L) - narg;  /* function index */
   lua_pushcfunction(L, traceback);  /* push traceback function */
   lua_insert(L, base);  /* put it under chunk and args */
-  signal(SIGINT, laction);
+  sig_catch(SIGINT, laction);
   status = lua_pcall(L, narg, (clear ? 0 : LUA_MULTRET), base);
-  signal(SIGINT, SIG_DFL);
+  sig_catch(SIGINT, SIG_DFL);
   lua_remove(L, base);  /* remove traceback function */
   /* force a complete garbage collection in case of errors */
   if (status != 0) lua_gc(L, LUA_GCCOLLECT, 0);
